### DISCUSSION TITLE
Phalcon Portfolio Projects on subdirs

### DISCUSSION THREAD

**Discussion by Bjunk** [phalcon/cphalcon#15274] [Q&A]
Hello Gents,

I have some working Phalcon projects and I would like to put all together as Portfolio, so I'm having some issues where Phalcon projects are not working outside the DocumentRoot

DocumentRoot /var/www/demo/
<Directory "/var/www/demo">
         Options All
         AllowOverride All
        Order allow,deny
        Allow from all
</Directory>

On DocumentRoot I have many subdirectories like demo01, demo02 with different Phalcon Apps but when I tried to reached it I received an 404 error. If I move each one Phalcon Apps to the DocumentRoot are working well.

(Idea: https://demo/demo01/, https://demo/demo02/ .. )
/
/demo01/
       + app
       + public
       + ..
/demo02/
       + app
       + public
       + ..
/demo03/
      ..

I added on the .htaccess files a RewriteBase /demo01/ but It's not working. I reviewed the Loglevel debug on my apache server to check the mod_rewrite but I don't see nothing special.

My questions: is it possible to have something like this? I will appreciate your feedback and recommendations. 

Thanks
Pablo

**Comment by quasipickle** [comment]
This is absolutely possible.  My guess is each app is configured using the document root as a basis, rather than the current directory.

For example, maybe you have something like:

```php
$base = $_SERVER['DOCUMENT_ROOT']; // This will be the same no matter where you move your project
$autoload = [
    'Model'=>$base.'/models';
];
$Loader = new \Phalcon\Loader();
$Loader->registerNamespaces($autoload);
$Loader->register();
```
instead of:
```php
$base = __DIR__; // This will change when you move your project
$autoload = [
    'Model'=>$base.'/models';
];
$Loader = new \Phalcon\Loader();
$Loader->registerNamespaces($autoload);
$Loader->register();
```

**Comment by Bjunk** [comment]
Thank Dylan for the reply and suggestion. My configuration is below: (BTW: I'm using Phalcon 3.4.5)

Config.php file:
defined('BASE_PATH') || define('BASE_PATH', getenv('BASE_PATH') ?: realpath(dirname(__FILE__) . '/../..'));
defined('APP_PATH') || define('APP_PATH', BASE_PATH . '/app');
'baseUri'        => '/demo01/' (following Invo Tutorial)

It's using realpath(dirname(__FILE__) . '/../..') on the config file.

Using out-of-the-box .htaccess files by dev-tools. I received an 404 response.

If I tried for example https://demo/demo01/ is not working (404 response), but https://demo/demo01/js/scripts.js is working. Any file on public I can reached except index.php. Apache2 is using a default configuration.

For me it's clear that it's related with baseUri or where the App is trying to find the documentRoot

I will appreciate any advice.

**Reply by quasipickle** [reply]
Yeah, the default `.htaccess` I believe doesn't redirect if the file is an actual file (ie: a javascript file).

Can you post your `config.php` and `bootstrap.php` files?  Being careful to remove any credentials or private information.

**Comment by Bjunk** [comment]
Config File: 

<?php
/*
 * Modified: prepend directory path of current file, because of this file own different ENV under between Apache and command line.
 * NOTE: please remove this comment.
 */
defined('BASE_PATH') || define('BASE_PATH', getenv('BASE_PATH') ?: realpath(dirname(__FILE__) . '/../..'));
defined('APP_PATH') || define('APP_PATH', BASE_PATH . '/app');

return new \Phalcon\Config([
    'database' => [
        'adapter'     => 'Mysql',
        'host'        => '', //host_db
        'username'    => '',// user
        'password'    => '',// password
        'dbname'      => '', //dbname
        'charset'     => 'utf8',
    ],
    'application' => [
        'appDir'         => APP_PATH . '/',
        'controllersDir' => APP_PATH . '/controllers/',
        'modelsDir'      => APP_PATH . '/models/',
        'migrationsDir'  => APP_PATH . '/migrations/',
        'viewsDir'       => APP_PATH . '/views/',
        'pluginsDir'     => APP_PATH . '/plugins/',
        'libraryDir'     => APP_PATH . '/library/',
        'cacheDir'       => BASE_PATH . '/cache/',
        'baseUri'        => '/',
    ]
]);

Service.php
<?php

use Phalcon\Mvc\Router;
use Phalcon\Mvc\View;
use Phalcon\Mvc\View\Engine\Php as PhpEngine;
use Phalcon\Mvc\Url as UrlResolver;
use Phalcon\Mvc\View\Engine\Volt as VoltEngine;
use Phalcon\Mvc\Model\Metadata\Memory as MetaDataAdapter;
use Phalcon\Session\Adapter\Files as SessionAdapter;
use Phalcon\Flash\Direct as Flash;
use \Phalcon\Mvc\Dispatcher as PhDispatcher;

/**
 * Shared configuration service
 */
$di->setShared('config', function () {
    return include APP_PATH . "/config/config.php";
});

/**
 * The URL component is used to generate all kind of urls in the application
 */
$di->setShared('url', function () {
    $config = $this->getConfig();

    $url = new UrlResolver();
    $url->setBaseUri($config->application->baseUri);

    return $url;
});

/**
 * Setting up the view component
 */
$di->setShared('view', function () {
    $config = $this->getConfig();

    $view = new View();
    $view->setDI($this);
    $view->setViewsDir($config->application->viewsDir);

    $view->registerEngines([
        '.volt' => function ($view) {
            $config = $this->getConfig();

            $volt = new VoltEngine($view, $this);

            $volt->setOptions([
                'compiledPath' => $config->application->cacheDir,
                'compiledSeparator' => '_'
            ]);

            return $volt;
        },
        '.phtml' => PhpEngine::class

    ]);

    return $view;
});

/**
 * Database connection is created based in the parameters defined in the configuration file
 */
$di->setShared('db', function () {
    $config = $this->getConfig();

    $class = 'Phalcon\Db\Adapter\Pdo\\' . $config->database->adapter;
    $params = [
        'host'     => $config->database->host,
        'username' => $config->database->username,
        'password' => $config->database->password,
        'dbname'   => $config->database->dbname,
        'charset'  => $config->database->charset
    ];

    if ($config->database->adapter == 'Postgresql') {
       unset($params['charset']);
    }

    $connection = new $class($params);

    return $connection;
});


/**
 * If the configuration specify the use of metadata adapter use it or use memory otherwise
 */
$di->setShared('modelsMetadata', function () {
    return new MetaDataAdapter();
});

/**
 * Register the session flash service with the Twitter Bootstrap classes
 */
$di->set('flash', function () {
    return new Flash([
        'error'   => 'alert alert-danger',
        'success' => 'alert alert-success',
        'notice'  => 'alert alert-info',
        'warning' => 'alert alert-warning'
    ]);
});

/**
 * Start the session the first time some component request the session service
 */
$di->setShared('session', function () {
    $session = new SessionAdapter();
    $session->start();

    return $session;
});

/**
 * Register router
 */
$di->setShared('router', function () {
    $router = new Router();
    $router->setUriSource(
        Router::URI_SOURCE_SERVER_REQUEST_URI
    );

    return $router;
});

/*
 * Error 404
 */

$di->set(
    'dispatcher',
    function() use ($di) {

        $evManager = $di->getShared('eventsManager');

        $evManager->attach(
            "dispatch:beforeException",
            function($event, $dispatcher, $exception)
            {
                switch ($exception->getCode()) {
                    case PhDispatcher::EXCEPTION_HANDLER_NOT_FOUND:
                    case PhDispatcher::EXCEPTION_ACTION_NOT_FOUND:
                       $dispatcher->forward(
                            array(
                                'controller' => 'error',
                                'action'     => 'show404',
                            )
                        );
                        return false; 
                                    break;
                    default:
                        return false;
                        break;
                }
            }
        );
        $dispatcher = new PhDispatcher();
        $dispatcher->setEventsManager($evManager);
        return $dispatcher;
    },
    true
);

**Reply by quasipickle** [reply]
Please wrap it in three backticks so it can be syntax highlighted and spaced.

**Comment by Bjunk** [comment]
Config.php
```
<?php
/*
 * Modified: prepend directory path of current file, because of this file own different ENV under between Apache and command line.
 * NOTE: please remove this comment.
 */
defined('BASE_PATH') || define('BASE_PATH', getenv('BASE_PATH') ?: realpath(dirname(__FILE__) . '/../..'));
defined('APP_PATH') || define('APP_PATH', BASE_PATH . '/app');

return new \Phalcon\Config([
    'database' => [
        'adapter'     => 'Mysql',
        'host'        => '',//host
        'username'    => '',//user
        'password'    => '',//password
        'dbname'      => '',//dbname
        'charset'     => 'utf8',
    ],
    'application' => [
        'appDir'         => APP_PATH . '/',
        'controllersDir' => APP_PATH . '/controllers/',
        'modelsDir'      => APP_PATH . '/models/',
        'migrationsDir'  => APP_PATH . '/migrations/',
        'viewsDir'       => APP_PATH . '/views/',
        'pluginsDir'     => APP_PATH . '/plugins/',
        'libraryDir'     => APP_PATH . '/library/',
        'cacheDir'       => BASE_PATH . '/cache/',
        'baseUri'        => '/',
    ]
]);
```

Service.php
```
<?php

use Phalcon\Mvc\Router;
use Phalcon\Mvc\View;
use Phalcon\Mvc\View\Engine\Php as PhpEngine;
use Phalcon\Mvc\Url as UrlResolver;
use Phalcon\Mvc\View\Engine\Volt as VoltEngine;
use Phalcon\Mvc\Model\Metadata\Memory as MetaDataAdapter;
use Phalcon\Session\Adapter\Files as SessionAdapter;
use Phalcon\Flash\Direct as Flash;
use \Phalcon\Mvc\Dispatcher as PhDispatcher;

/**
 * Shared configuration service
 */
$di->setShared('config', function () {
    return include APP_PATH . "/config/config.php";
});

/**
 * The URL component is used to generate all kind of urls in the application
 */
$di->setShared('url', function () {
    $config = $this->getConfig();

    $url = new UrlResolver();
    $url->setBaseUri($config->application->baseUri);

    return $url;
});

/**
 * Setting up the view component
 */
$di->setShared('view', function () {
    $config = $this->getConfig();

    $view = new View();
    $view->setDI($this);
    $view->setViewsDir($config->application->viewsDir);

    $view->registerEngines([
        '.volt' => function ($view) {
            $config = $this->getConfig();

            $volt = new VoltEngine($view, $this);

            $volt->setOptions([
                'compiledPath' => $config->application->cacheDir,
                'compiledSeparator' => '_'
            ]);

            return $volt;
        },
        '.phtml' => PhpEngine::class

    ]);

    return $view;
});

/**
 * Database connection is created based in the parameters defined in the configuration file
 */
$di->setShared('db', function () {
    $config = $this->getConfig();

    $class = 'Phalcon\Db\Adapter\Pdo\\' . $config->database->adapter;
    $params = [
        'host'     => $config->database->host,
        'username' => $config->database->username,
        'password' => $config->database->password,
        'dbname'   => $config->database->dbname,
        'charset'  => $config->database->charset
    ];

    if ($config->database->adapter == 'Postgresql') {
        unset($params['charset']);
    }

    $connection = new $class($params);

    return $connection;
});


/**
 * If the configuration specify the use of metadata adapter use it or use memory otherwise
 */
$di->setShared('modelsMetadata', function () {
    return new MetaDataAdapter();
});

/**
 * Register the session flash service with the Twitter Bootstrap classes
 */
$di->set('flash', function () {
    return new Flash([
        'error'   => 'alert alert-danger',
        'success' => 'alert alert-success',
        'notice'  => 'alert alert-info',
        'warning' => 'alert alert-warning'
    ]);
});

/**
 * Start the session the first time some component request the session service
 */
$di->setShared('session', function () {
    $session = new SessionAdapter();
    $session->start();

    return $session;
});

/**
 * Register router
 */
$di->setShared('router', function () {
    $router = new Router();
    $router->setUriSource(
        Router::URI_SOURCE_SERVER_REQUEST_URI
    );

    return $router;
});

/*
 * Error 404
 */

$di->set(
    'dispatcher',
    function() use ($di) {

        $evManager = $di->getShared('eventsManager');

        $evManager->attach(
            "dispatch:beforeException",
            function($event, $dispatcher, $exception)
            {
                switch ($exception->getCode()) {
                    case PhDispatcher::EXCEPTION_HANDLER_NOT_FOUND:
                    case PhDispatcher::EXCEPTION_ACTION_NOT_FOUND:
                       $dispatcher->forward(
                            array(
                                'controller' => 'error',
                                'action'     => 'show404',
                            )
                        );
                        return false; 
                                    break;
                    default:
                         /* $dispatcher->forward(array(
                                'controller' => 'error',
                                'action' => 'show500',
                        )); */
                        return false;
                        break;
                                'controller' => 'error',
                                'action'     => 'show404',
                            )
                        );
                        return false; 
                                    break;
                    default:
                         /* $dispatcher->forward(array(
                                'controller' => 'error',
                                'action' => 'show500',
                        )); */
                        return false;
                        break;
                }
            }
        );
        $dispatcher = new PhDispatcher();
        $dispatcher->setEventsManager($evManager);
        return $dispatcher;
    },
    true
);
```

**Reply by quasipickle** [reply]
You're not getting anything in the Apache logs because as far as Apache knows, the URL is being handled.

As for why you're getting the 404, it's because Phalcon can't find a controller or controller action to match the URL.  You're not setting up the autoloader anywhere, so I'm guessing Phalcon doesn't know where to look.  Why don't you detach that `$evManager` and see if you get any helpful exceptions/errors from Phalcon.

**Comment by Bjunk** [comment]
Thanks Dylan. I dettached the $evManager and now I received this:

Demo01Controller handler class cannot be loaded

```
#0 [internal function]: Phalcon\Mvc\Dispatcher->_throwDispatchException('Demo01Controlle...', 2)
#1 [internal function]: Phalcon\Dispatcher->dispatch()
#2 /var/www/demo.dxs.cloud/demo01/public/index.php(42): Phalcon\Mvc\Application->handle()
#3 {main}
```

**Reply by quasipickle** [reply]
That's looking more and more like an autoloader not being set up.

You'll want to do something like:

```php
$Config = $di->getConfig();

// This sets up Phalcon to know where to look, based on namespaces used
$Loader = new \Phalcon\Loader();
$Loader->registerNamespaces([
    'Controller' => $Config->application->controllersDir, // Assuming your controller namespace is "Controller"
    'Model'      => $Config->application->modelsDir
]);
$Loader->register();

// This sets up the dispatcher to use the "Controller" namespace by default
$di->setShared('dispatcher',function(){

    // Your events manager stuff goes here

    $Dispatcher = new \Phalcon\Mvc\Dispatcher();
    $Dispatcher->setDefaultNamespace('Controller');
    $Dispatcher->setEventsManager($EM);

    return $Dispatcher;
});
```

And just to be 100% sure, why `echo` out `$Config->application->controllersDir` - just to see it's definitely what you want.

**Comment by Bjunk** [comment]
I configured in config.php 

```
'baseUri'        => '/demo01/'
```

and nothing changed.

It's trying to find the directory name as Controller clearly

**Reply by quasipickle** [reply]
I don't believe the `baseUri` has anything to do with the autoloader.  I'm writing a response to your earlier post - finished shortly.

**Comment by Bjunk** [comment]
I made the change that you suggested, but continue the same issue. Phalcon is trying to find everything on DocumentRoot. I thought that baseUri will help based on "invo tutorial" for Phalcon 3.4

```
Controllers\Demo01Controller handler class cannot be loaded

#0 [internal function]: Phalcon\Mvc\Dispatcher->_throwDispatchException('Controllers\\Dem...', 2)
#1 [internal function]: Phalcon\Dispatcher->dispatch()
#2 /var/www/demo.dxs.cloud/demo01/public/index.php(42): Phalcon\Mvc\Application->handle()
#3 {main}
```

All my files are the default files created by dev-tools. No changes, except by the evManager.

**Comment by Bjunk** [comment]
Thank Dylan for all your help. I'm using a out-of-the-box dev-tools configuration files and happen the same issue using files outside DocumentRoot

**Reply by quasipickle** [reply]
Is your controller actually called `Demo01Controller`?  Or are you renaming it for posting here?  If renaming, check your capitalization.  For example: `AdminProductController` won't work, but `AdminproductController` will.

You can attach an event listener to the loader so it outputs every path it checks.  Maybe that will show you where Phalcon is looking.  You can see an example here: https://docs.phalcon.io/3.4/en/loader#autoloading-events

One other thing you can try to hopefully get some more details about the error is to enable the Phalcon debugger.  At the top of your index.php file, put these lines:

```php
$Debug = new \Phalcon\Debug();
$Debug->listen();
```

Then if any exceptions get thrown you get a nice UI and a complete trace.
